tests/test_gradient.py

```diff
@@ -1081,3 +1083,49 @@ def test_jacobian_disconnected_inputs():
     func_s = pytensor.function([s2], jacobian_s)
     val = np.array(1.0).astype(pytensor.config.floatX)
     assert np.allclose(func_s(val), np.zeros(1))
+
+
+def test_hessp():
+    x = vector("x", dtype="float64")
+    rosen = (100 * (x[1:] - x[:-1] ** 2) ** 2 + (1 - x[:-1]) ** 2).sum()
+    p = vector("p")
+
+    rosen_hess_prod_pt = hessian_prod(rosen, wrt=x, p=p)
+
+    x_test = 0.1 * np.arange(9)
+    p_test = 0.5 * np.arange(9)
+    np.testing.assert_allclose(
+        rosen_hess_prod_pt.eval({x: x_test, p: p_test}),
+        rosen_hess_prod(x_test, p_test),
+    )
+
+
+def test_hessp_example():
+    import numpy as np
+    from scipy.optimize import minimize
+
+    from pytensor import function
+    from pytensor.gradient import hessian_prod, jacobian
+    from pytensor.tensor import vector
+
+    x = vector("x")
+    p = vector("p")
+
+    rosen = (100 * (x[1:] - x[:-1] ** 2) ** 2 + (1 - x[:-1]) ** 2).sum()
+    rosen_hessp = hessian_prod(rosen, x, p)
+    rosen_jac = jacobian(rosen, x)
+
+    rosen_fn = function([x], rosen)
+    rosen_jac_fn = function([x], rosen_jac)
+    rosen_hessp_fn = function([x, p], rosen_hessp)
+    x0 = np.array([1.3, 0.7, 0.8, 1.9, 1.2])
+    res = minimize(
+        rosen_fn,
+        x0,
+        method="Newton-CG",
+        jac=rosen_jac_fn,
+        hessp=rosen_hessp_fn,
+        options={"xatol": 1e-8},
+    )
+    assert res.success
+    np.testing.assert_allclose(res.x, np.ones_like(x0), atol=1e-3)
```

Comment:
@aseyboldt Any idea why Scipy stops at a solution with higher atol error? It seems to work better in their example: https://docs.scipy.org/doc/scipy/tutorial/optimize.html#hessian-product-example 

Is our jac / hessp code crap?